File: File.cpp
//  Copyright (C) 2001, 2002 Matthew Landauer. All Rights Reserved.
//  
//  This program is free software; you can redistribute it and/or modify it
//  under the terms of version 2 of the GNU General Public License as
//  published by the Free Software Foundation.
//
//  This program is distributed in the hope that it would be useful, but
//  WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  Further, any
//  license provided herein, whether implied or otherwise, is limited to
//  this program in accordance with the express provisions of the GNU
//  General Public License.  Patent licenses, if any, provided herein do not
//  apply to combinations of this program with other product or programs, or
//  any other product whatsoever.  This program is distributed without any
//  warranty that the program is delivered free of the rightful claim of any
//  third person by way of infringement or the like.  See the GNU General
//  Public License for more details.
//
//  You should have received a copy of the GNU General Public License along
//  with this program; if not, write the Free Software Foundation, Inc., 59
//  Temple Place - Suite 330, Boston MA 02111-1307, USA.
//
// $Id$

#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <unistd.h>
#include <iostream>

#include "File.h"

namespace Sp {

File::File(const Path &path) : FsObject(path), fileOpen(false)
{
}

bool File::valid() const
{
	struct stat fileStat;
	int ret = lstat(path().fullName().c_str(), &fileStat);
	if (ret == 0)
		return(S_ISREG(fileStat.st_mode));
	else
		return false;
}

/*!
  \todo Opens for read only at the moment
*/
void File::open()
{
	fd = ::open(path().fullName().c_str(), O_RDONLY);
	// TEMPORARY HACK
	if (fd == -1)
		std::cerr << "Error opening file " << path().fullName().c_str() << std::endl;
	else
		fileOpen = true;
}

Size File::size() const
{
	return (Size::Bytes(sizeBytes()));
}

long int File::sizeBytes() const
{
	struct stat fileStat;
	int ret = lstat(path().fullName().c_str(), &fileStat);
  assert(ret == 0);
	return fileStat.st_size;
}

void File::close()
{
	::close(fd);
	fileOpen = false;
}

unsigned long int File::read(void *buf, unsigned long int count) const
{
	return (::read(fd, buf, count));
}

void File::seek(unsigned long int pos) const
{
	off_t offset = lseek(fd, pos, SEEK_SET);
	assert(offset != -1);
}

void File::seekForward(unsigned long int pos) const
{
	off_t offset = lseek(fd, pos, SEEK_CUR);
	assert(offset != -1);
}

unsigned char File::readChar() const
{
	unsigned char value;
	unsigned long no = read(&value, 1);
	assert(no == 1);
	return (value);
}

unsigned short File::readShort(const int &endian) const
{
	unsigned short value;
	unsigned char temp[2];
	unsigned long no = read(temp, 2);
	assert (no == 2);
	
	// If small endian
	if (endian == 0)
		value = (temp[0]<<0) + (temp[1]<<8);
	else
		value = (temp[0]<<8) + (temp[1]<<0);
	return (value);
}

unsigned long File::readLong(const int &endian) const
{
	unsigned char temp[4];
	unsigned long no = read(temp, 4);
	assert (no == 4);
	
	unsigned long value;
	if (endian == 0)
		value = (temp[0]<<0) + (temp[1]<<8) + (temp[2]<<16) + (temp[3]<<24);
	else
		value = (temp[0]<<24) + (temp[1]<<16) + (temp[2]<<8) + (temp[3]<<0);
	return (value);
}

}

File: app/models/earth/file.rb
# Copyright (C) 2007 Rising Sun Pictures and Matthew Landauer
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

module Earth
  class File < ActiveRecord::Base
    belongs_to :directory
    composed_of :user, :mapping => [%w(uid uid)]
    
    Stat = Struct.new(:mtime, :size, :blocks, :uid, :gid)
    class Stat
      def ==(s)
        mtime == s.mtime && size == s.size && blocks == s.blocks && uid == s.uid && gid == s.gid
      end
    end
    
    # Convenience method for setting all the fields associated with stat in one hit
    def stat=(stat)
      self.modified = stat.mtime.utc
      self.bytes = stat.size
      self.blocks = stat.blocks
      self.uid = stat.uid
      self.gid = stat.gid
    end
    
    # Returns a "fake" Stat object with some of the same information as File::Stat
    def stat
      Stat.new(modified, bytes, blocks, uid, gid)
    end
    
    def size
      Size.new(bytes, blocks, 1)
    end
    
    def empty?
      bytes == 0
    end
    
    def hidden?
      name && name[0] == ?.
    end
    
    def self.filter(files,options={})
      show_hidden = options[:show_hidden]
      
      any_empty  = false
      any_hidden = false
      
      filtered = files.map do |file|
        any_empty  = true if file.empty?
        any_hidden = true if file.hidden?
        
        file if show_hidden || !file.hidden?
      end
      
      [filtered.compact, any_empty, any_hidden]
    end

    def path
      File.join(directory.path, name)
    end
    
    def self.build_filter_conditions(params)
      filter_filename = params[:filter_filename]
      filter_filename = "*" if filter_filename.blank?
      
      filter_user = params[:filter_user]
      filter_uid  = filter_user.blank? ? nil : User.find_by_name(filter_user).uid

      if filter_uid
        ["files.name LIKE ? AND files.uid = ?", filter_filename.tr('*', '%'), filter_uid]
      elsif filter_filename != '*'
        ["files.name LIKE ?", filter_filename.tr('*', '%')]
      end
    end
    
    def self.with_filter(params = {})
      filter_conditions = build_filter_conditions(params)

      Thread.current[:with_filtering] = filter_conditions
      
      Earth::File.with_scope(:find => {:conditions => filter_conditions}) do
        begin
          yield
        ensure
          Thread.current[:with_filtering] = nil
        end
      end
    end
  end
end
